File: go/algorithms/binary_search.go
package algorithms

import "cmp"

// BinarySearch searches for target in a sorted slice.
// Returns the index if found, -1 otherwise.
// Time: O(log n)
// Space: O(1)
func BinarySearch[T cmp.Ordered](arr []T, target T) int {
	left, right := 0, len(arr)-1

	for left <= right {
		mid := left + (right-left)/2

		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return -1
}

// LowerBound returns the index of the first element not less than target.
func LowerBound[T cmp.Ordered](arr []T, target T) int {
	left, right := 0, len(arr)

	for left < right {
		mid := left + (right-left)/2
		if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid
		}
	}

	return left
}

// UpperBound returns the index of the first element greater than target.
func UpperBound[T cmp.Ordered](arr []T, target T) int {
	left, right := 0, len(arr)

	for left < right {
		mid := left + (right-left)/2
		if arr[mid] <= target {
			left = mid + 1
		} else {
			right = mid
		}
	}

	return left
}

File: kotlin/src/main/kotlin/algorithms/BinarySearch.kt
package algorithms

/**
 * Binary search algorithms.
 * Time: O(log n)
 * Space: O(1)
 */
object BinarySearch {
    /**
     * Find the index of target in a sorted list.
     * Returns -1 if not found.
     */
    fun <T : Comparable<T>> search(list: List<T>, target: T, comparator: Comparator<T>? = null): Int {
        var left = 0
        var right = list.lastIndex

        while (left <= right) {
            val mid = left + (right - left) / 2
            val cmp = compare(list[mid], target, comparator)

            when {
                cmp == 0 -> return mid
                cmp < 0 -> left = mid + 1
                else -> right = mid - 1
            }
        }

        return -1
    }

    /**
     * Find the leftmost index where target could be inserted to maintain sorted order.
     * Returns the index of the first element >= target.
     */
    fun <T : Comparable<T>> lowerBound(list: List<T>, target: T, comparator: Comparator<T>? = null): Int {
        var left = 0
        var right = list.size

        while (left < right) {
            val mid = left + (right - left) / 2
            if (compare(list[mid], target, comparator) < 0) {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return left
    }

    /**
     * Find the rightmost index where target could be inserted to maintain sorted order.
     * Returns the index of the first element > target.
     */
    fun <T : Comparable<T>> upperBound(list: List<T>, target: T, comparator: Comparator<T>? = null): Int {
        var left = 0
        var right = list.size

        while (left < right) {
            val mid = left + (right - left) / 2
            if (compare(list[mid], target, comparator) <= 0) {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return left
    }

    private fun <T : Comparable<T>> compare(a: T, b: T, comparator: Comparator<T>?): Int =
        comparator?.compare(a, b) ?: a.compareTo(b)
}
